src/lib/viewers/media/MediaBaseViewer.js

```diff
@@ -680,16 +707,66 @@ class MediaBaseViewer extends BaseViewer {
-        this.mediaEl.addEventListener('playing', this.playingHandler);
+        this.mediaEl.addEventListener('canplay', this.stopBufferFillTimer);
+        this.mediaEl.addEventListener('ended', this.mediaendHandler);
+        this.mediaEl.addEventListener('loadstart', this.startBufferFillTimer);
         this.mediaEl.addEventListener('pause', this.pauseHandler);
-        this.mediaEl.addEventListener('ended', this.resetPlayIcon);
+        this.mediaEl.addEventListener('playing', this.playingHandler);
+        this.mediaEl.addEventListener('progress', this.progressHandler);
         this.mediaEl.addEventListener('seeked', this.seekHandler);
-        this.mediaEl.addEventListener('ended', this.mediaendHandler);
+        this.mediaEl.addEventListener('timeupdate', this.setTimeCode);
+        this.mediaEl.addEventListener('volumechange', this.updateVolumeIcon);
     }
 
+    /**
+     * Removes event listeners to the media element
+     * @return {void}
+     */
+    removeEventListenersForMediaElement() {
+        this.mediaEl.removeEventListener('canplay', this.stopBufferFillTimer);
+        this.mediaEl.removeEventListener('ended', this.mediaendHandler);
+        this.mediaEl.removeEventListener('loadstart', this.startBufferFillTimer);
+        this.mediaEl.removeEventListener('pause', this.pauseHandler);
+        this.mediaEl.removeEventListener('playing', this.playingHandler);
+        this.mediaEl.removeEventListener('progress', this.progressHandler);
+        this.mediaEl.removeEventListener('seeked', this.seekHandler);
+        this.mediaEl.removeEventListener('timeupdate', this.setTimeCode);
+        this.mediaEl.removeEventListener('volumechange', this.updateVolumeIcon);
+    }
+
+    /**
+     * Callback from the 'loadstart' event from the media element. Triggers a timer to measure the initial buffer fill.
+     * @return {void}
+     */
+    startBufferFillTimer() {
+        const tag = Timer.createTag(this.options.file.id, MEDIA_METRIC.bufferFill);
+        Timer.start(tag);
+    }
+
+    /**
+     * Callback from the 'canplay' event from the media element. The first time this event is triggered we
+     * calculate the initial buffer fill time.
+     * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canplay_event}
+     * @emits MEDIA_METRIC_EVENTS.bufferFill
+     * @return {void}
+     */
+    stopBufferFillTimer() {
+        const tag = Timer.createTag(this.options.file.id, MEDIA_METRIC.bufferFill);
+        Timer.stop(tag);
+
+        // Only interested in the first event after 'loadstart' to determine the buffer fill
+        this.mediaEl.removeEventListener('canplay', this.stopBufferFillTimer);
+
+        this.processBufferFillMetric();
+    }
+
+    /**
+     * Abstract. Processes the buffer fill metric which represents the initial buffer time before playback begins
+     * @emits MEDIA_METRIC_EVENTS.bufferFill
+     * @return {void}
+     */
+    processBufferFillMetric() {}
```

Comment:
Will we be able to share this logic between Dash/MP4 viewers? If not, should they instead override the handlers and each add their own additional method calls?

Reply:
They should be able to share the logic, but right now I only want to enable it for DashViewer